File: 2024/Day4/1.py
def count_xmas(grid, word="XMAS"):
    # Get dimensions of the grid
    rows, cols = len(grid), len(grid[0])
    word_length = len(word)
    directions = [
        (0, 1),  # Right
        (1, 0),  # Down
        (1, 1),  # Diagonal down-right
        (1, -1), # Diagonal down-left
        (0, -1), # Left
        (-1, 0), # Up
        (-1, -1),# Diagonal up-left
        (-1, 1)  # Diagonal up-right
    ]

    def in_bounds(r, c):
        return 0 <= r < rows and 0 <= c < cols

    def match(r, c, dr, dc):
        for i in range(word_length):
            nr, nc = r + i * dr, c + i * dc
            if not in_bounds(nr, nc) or grid[nr][nc] != word[i]:
                return False
        return True

    count = 0
    for r in range(rows):
        for c in range(cols):
            for dr, dc in directions:
                if match(r, c, dr, dc):
                    count += 1
    return count

grid = list()

with open("input.txt", "r") as file:
  content = file.readlines()
  for line in content:
    grid.append(list(line.strip()))

# Count occurrences of "XMAS"
result = count_xmas(grid)
print(f"Total occurrences of 'XMAS': {result}")

File: 2024/Day4/1.swift
import Foundation

func countXMAS(grid: [[Character]], word: String = "XMAS") -> Int {
    let rows = grid.count
    let cols = grid[0].count
    let wordLength = word.count
    let directions = [
        (0, 1),   // Right
        (1, 0),   // Down
        (1, 1),   // Diagonal down-right
        (1, -1),  // Diagonal down-left
        (0, -1),  // Left
        (-1, 0),  // Up
        (-1, -1), // Diagonal up-left
        (-1, 1)   // Diagonal up-right
    ]

    func inBounds(row: Int, col: Int) -> Bool {
        return row >= 0 && row < rows && col >= 0 && col < cols
    }

    func match(row: Int, col: Int, dr: Int, dc: Int) -> Bool {
        for i in 0..<wordLength {
            let nr = row + i * dr
            let nc = col + i * dc
            if !inBounds(row: nr, col: nc) || grid[nr][nc] != word[word.index(word.startIndex, offsetBy: i)] {
                return false
            }
        }
        return true
    }

    var count = 0

    for row in 0..<rows {
        for col in 0..<cols {
            for (dr, dc) in directions {
                if match(row: row, col: col, dr: dr, dc: dc) {
                    count += 1
                }
            }
        }
    }

    return count
}

// Read input from file
let fileURL = URL(fileURLWithPath: "input.txt")
if let content = try? String(contentsOf: fileURL, encoding: .utf8) {
    let grid = content.split(separator: "\n").map { Array($0) }
    let result = countXMAS(grid: grid)
    print("Total occurrences of 'XMAS': \(result)")
} else {
    print("Error reading the file.")
}
